### PR TITLE
Fix PyQGIS QgsLineString constructor only accepts lists of QgsPoint, not QgsPointXY as indicated by the documentation

### LINKED ISSUE #43200: QgsLineString does not accept iterable[QgsPointXY] on costructor method
**Describe the bug**
QgsLineString does not accept iterable[QgsPointXY] on costructor method

It's not a big problem since we could generate a QgsPoint from a QgsPointXY, but the documentation (https://qgis.org/pyqgis/master/core/QgsLineString.html) suggests that QgsLineString should accept QgsPointXY on constructor: " New in version 3.2.
QgsLineString(points: Iterable[QgsPointXY]) Construct a linestring from list of points."

**How to Reproduce**

On console, type:
from qgis.core import QgsLineString
from qgis.core import QgsPointXY
p1 = QgsPointXY(1,1)
p2 = QgsPointXY(2,2)
p3 = QgsPointXY(4,0)
QgsLineString([p1,p2,p3])

**QGIS and OS versions**

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /></head><body>

QGIS version | 3.18.1-Zürich | QGIS code revision | 202f1bf7e5
-- | -- | -- | --
Compiled against Qt | 5.11.2 | Running against Qt | 5.11.2
Compiled against GDAL/OGR | 3.1.4 | Running against GDAL/OGR | 3.1.4
Compiled against GEOS | 3.8.1-CAPI-1.13.3 | Running against GEOS | 3.8.1-CAPI-1.13.3
Compiled against SQLite | 3.29.0 | Running against SQLite | 3.29.0
PostgreSQL Client Version | 11.5 | SpatiaLite Version | 4.3.0
QWT Version | 6.1.3 | QScintilla2 Version | 2.10.8
Compiled against PROJ | 6.3.2 | Running against PROJ | Rel. 6.3.2, May 1st, 2020
OS Version | Windows 10 (10.0)
Active python plugins | DsgTools; plugin_reloader; quick_map_services; db_manager; MetaSearch; processing; ferramentas_experimentais

</body></html>

### PR DESCRIPTION
**Pull request by nyalldawson**
Also add support for constructing QgsLineString using arrays of
arrays of floats, given that we're having to hand-roll sip conversion
code anyway!

Now the following is supported:

    line = QgsLineString([[1,2], [3,4], [5,6]])

which is much nicer and more "pythonic" then the explicit
QgsPoint/QgsPointXY sequences!

Fixes #43200

### CODE REVIEW

File: src/core/geometry/qgslinestring.h [hunk 1]
```diff
@@ -58,6 +59,204 @@ class CORE_EXPORT QgsLineString: public QgsCurve
      */
     QgsLineString( const QVector<QgsPoint> &points ) SIP_HOLDGIL;
```
Comment (3nids) [inline src/core/geometry/qgslinestring.h:60]: this method is not exposed to sipify no?
```suggestion
    QgsLineString( const QVector<QgsPoint> &points );
```

File: src/core/geometry/qgslinestring.h [hunk 2]
```diff
@@ -58,6 +59,204 @@ class CORE_EXPORT QgsLineString: public QgsCurve
...
 
+    /**
+     * Construct a linestring from list of points.
+     * This constructor is more efficient then calling setPoints()
+     * or repeatedly calling addVertex()
+     * \since QGIS 3.0
+     */
+    QgsLineString( const QVector<QgsPointXY> &points ) SIP_HOLDGIL;
```
Comment (3nids) [inline src/core/geometry/qgslinestring.h:68]:
```suggestion
    QgsLineString( const QVector<QgsPointXY> &points );
```

File: src/core/geometry/qgslinestring.h [hunk 3]
```diff
@@ -58,6 +59,204 @@ class CORE_EXPORT QgsLineString: public QgsCurve
...
+#else
+
+    /**
+     * Construct a linestring from a sequence of points (QgsPoint objects, QgsPointXY objects, or sequences of float values).
+     *
+     * The linestring Z and M type will be set based on the type of the first point in the sequence.
+     *
+     * \since QGIS 3.20
+     */
+    QgsLineString( SIP_PYOBJECT points SIP_TYPEHINT( Sequence[Union[QgsPoint, QgsPointXY, Sequence[float]]] ) ) [( const QVector<double> &x, const QVector<double> &y, const QVector<double> &z = QVector<double>(), const QVector<double> &m = QVector<double>(), bool is25DType = false )];
```
Comment (3nids) [inline src/core/geometry/qgslinestring.h:78]: is this one missing `SIP_HOLDGIL` ?

**Review by 3nids — APPROVED** [review]
nice!

File: src/core/geometry/qgslinestring.h (hunk 3 above)
Comment (nyalldawson) [reply src/core/geometry/qgslinestring.h:78]: good catch, thanks!

**Pull request merged**